### ISSUE TITLE
FinalSteps/Queries other than Select missing getQuery method

### ISSUE THREAD

**Issue by nkoterba** [jOOQ/jOOQ#4899]
We are using a "builder"-like pattern to assemble some of our Jooq queries.  To do so, we need to get access to the underlying explicit `Query` "interfaces" being used.

For example, a select query is a `SelectQuery<Record>` while an update query is a `UpdateQuery<Record>`.

Unfortunately, only `select` seems to provide a means to get the underlying query object.  The `SelectFinalStep` interface has a `getQuery()` method that returns the actual `SelectQuery<R>`.

https://github.com/jOOQ/jOOQ/blob/94f0bb2f69895ad6bba8d6353efc33f77d85b358/jOOQ/src/main/java/org/jooq/SelectFinalStep.java#L91

Here is the concrete implementation:
https://github.com/jOOQ/jOOQ/blob/94f0bb2f69895ad6bba8d6353efc33f77d85b358/jOOQ/src/main/java/org/jooq/impl/SelectImpl.java#L225-L228

None of the other **CUD** Final steps or concrete implementations seem to expose or support this `getQuery()` method which would be incredibly useful in our scenario.

Is this an intentional omission?  Or just some missing functionality?

We need to call `addConditions` on our "query" object and without getting the more explicit query types (e.g. SelectQuery, DeleteQuery, UpdateQuery) the `addConditions` method is not accessible.

**Comment by lukaseder** [comment]
I'm marking this as a duplicate of #2198

> Is this an intentional omission?

Yes

> Or just some missing functionality?

Yes :-)

`SelectFinalStep.getQuery()` is something that we regret today. The method is not on the right interface, and the behaviour is not as useful as it could be. The DSL API `XXXStep` and the underlying model API (`XXXQuery`) will be re-designed in #2198.

I understand that the status quo is limiting for users who like to build most of their queries with the DSL API, and then patch only parts with the model API. Right now, however, I suggest you:

- Either use the model API from the beginning ( `DSLContext.selectQuery()`)
- Or, access the underlying `XXXQuery` type via reflection, or by putting a helper class in the `org.jooq.impl` package.

**Comment by nkoterba** [comment]
First thanks for the quick and informative reply. :+1: 

> Either use the model API from the beginning ( DSLContext.selectQuery())

So my understanding with this strategy is that we'd need to re-write all our queries to use the model API (e.g. addFrom, addSelect, addConditions, etc.).  This strategy moves us away from the nice DSL syntax and SQL-convention naming (where clauses, etc.).

**--> LOW desire**

> Or, access the underlying XXXQuery type via reflection

Here's a snippet of code we verified works in our scenario:
```java
// Get the original query
final org.jooq.Query q = queryGenerator.generateQuery(db);

try {
	Field f = q.getClass().getSuperclass().getDeclaredField("delegate");
	f.setAccessible(true); // Needed to prevent IllegalAccessException below
	SelectQuery selectQuery = (SelectQuery)f.get(q);
	selectQuery.addConditions(getDefaultSelectConditions());
} catch (SecurityException | NoSuchFieldException | IllegalAccessException e) {
	e.printStackTrace();
}

// return query with added 'where' conditions
return q;
```

**-> Wondering about performance impacts of using Reflection on every SQL call we make...**

> or by putting a helper class in the org.jooq.impl package.

We also explored this approach and came up with something like this (*notice the class is in the org.jooq.impl package although it resides in our src folder hierarchy*):
```java
package org.jooq.impl;

import org.jooq.Condition;
import org.jooq.Query;

import java.util.Collection;

public class JooqHackyFix {
	public static void addConditionsToQuery(Query q, Collection<Condition> conditions) {
		if (q instanceof SelectImpl)
			((SelectImpl)q).where(conditions);
		else if (q instanceof DeleteImpl)
			((DeleteImpl)q).where(conditions);
		else if (q instanceof UpdateImpl)
			((UpdateImpl)q).where(conditions);
	}
}
```

**-> Is probably more performant since a simple cast is made vs using reflection.  However, requires a "misplaced" class in our package so may fail a 'checkstyle' test that class' packages match their path.**

Both of the last two should be relatively easy to "update" or remove once the new DSL/Model API re-design is complete.

**Comment by lukaseder** [comment]
I certainly understand the low desire to move towards the model API - it isn't as easy to use as the DSL API at all.

I think that the `JooqHackyFix` might be the most optimal solution for the time being, as it doesn't depend as much on internals as the reflection fix. The reflection fix shouldn't be too bad from a performance perspective. If you're concerned, you could still cache the accessible field reference. The reflection lookup takes much longer than the actual usage.
